Adding the generic_package.sh script file

Shwetha-Acharya opened the pull request:
This script file is used to build the rpms for
Debian and Ubuntu versions.

We can run this script on a ubuntu machine.
We create a chroot environment using pbuilder create.
This gives us the environment for both ubuntu and debian.
On this we build, sign and push the packages.

We update the changelogs and push them to the glusterfs-debian
repo as well.

Signed-off-by: Sheetal Pamecha <spamecha@redhat.com>
Signed-off-by: hari gowtham <hgowtham@redhat.com>
Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
Adding the generic_package.sh script file

This script file is used to build the rpms for
Debian and Ubuntu versions.

We can run this script on a ubuntu machine.
We create a chroot environment using pbuilder create.
This gives us the environment for both ubuntu and debian.
On this we build, sign and push the packages.

We update the changelogs and push them to the glusterfs-debian
repo as well.

Signed-off-by: Sheetal Pamecha <spamecha@redhat.com>
Signed-off-by: hari gowtham <hgowtham@redhat.com>
Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
Adding package job and script.

This helps in automating the packaging task of Gluster for
Debian and Ubuntu.

Change-Id: Icdbc718f0a3f020715d959486d01e962f89fa80f
Signed-off-by: hari gowtham <hgowtham@redhat.com>
Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
Adding package job and script.

This helps in automating the packaging task of Gluster for
Debian and Ubuntu.

Change-Id: Icdbc718f0a3f020715d959486d01e962f89fa80f
Signed-off-by: hari gowtham <hgowtham@redhat.com>
Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
Adding package job and script.

This helps in automating the packaging task of Gluster for
Debian and Ubuntu.

Change-Id: Icdbc718f0a3f020715d959486d01e962f89fa80f
Signed-off-by: hari gowtham <hgowtham@redhat.com>
Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

mscherer commented on build-gluster-org/jobs/debian-ubuntu-package.yml:
```diff
@@ -0,0 +1,58 @@
+    - string:
+        default:
+        description: Release number for the package to be built against.
+          Leave it empty if you are building above series 5.
```
Mhh, what is series 5, you mean version 5 branch ? is it still supported, or can we drp it ?

Shwetha-Acharya replied:
We can drop them off.

mscherer replied:
So, the comment is still here, what is series 5 ? Do you mean "release 5" ? (and if we can drop, should it be removed from the description ?)

mscherer commented on build-gluster-org/scripts/debian-ubuntu-package.sh:
```diff
@@ -0,0 +1,75 @@
+if [ $# -eq 5 ]; then
+        echo "building everything"
+        echo "packing debian distribution"
+        ../scripts/generic_package.sh debian stretch $SERIES $VERSION $RELEASE $LATEST_SERIES $LATEST_VERSION
```
I think using ../scripts will not work. This will depend on the working directory jenkins use, and I suspect this will not be /opt/qa. better use the absolute path.

Also, I think a for loop will be easier to read.

Shwetha-Acharya replied:
done

mscherer commented on build-gluster-org/scripts/debian-ubuntu-package.sh:
```diff
@@ -0,0 +1,75 @@
+        ../scripts/generic_package.sh ubuntu focal $SERIES $VERSION $RELEASE
+    elif [ "$OS" == "debian" ]; then
+        echo "packing debian alone"
+        if [ "$FLAVOR" == "stretch" ] || [ "$FLAVOR" == "9" ]; then
```
I think it would be cleaner to have 1 single if with the 6 possibles values than 3 separate ones.

Shwetha-Acharya replied:
can you please elaborate this comment

mscherer replied:
Well, here we have 3 lines with if, they all run the same script with the same argument, the only difference is the echo. This seems harder to read while :

case $FLAVOR in
  "stretch" | "9" | "buster" | "10" | "bullseye" | "11") 
      ../scripts/generic_package.sh $OS $FLAVOR $SERIES $VERSION $RELEASE $LATEST_SERIES $LATEST_VERSION 

is shorter

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+tar czf ~/${os}-${flavor}-Glusterfs-${version}/${flavor}-apt-amd64-${version}.tgz pool/ dists/
+
+echo "Pushing Changelog changes.."
+git push origin ${flavor}-${series}-local:${flavor}-glusterfs-${series}
```
This is not going to work if the clone is done over HTTP

Shwetha-Acharya replied:
addressed

mscherer replied:
I am sorry, but I still see https for the clone. I was not precise enough, I want to point that we can't push over http, including https. And if we need to push a new changelog, we need a ssh key.

Shwetha-Acharya replied:
clone is now changed to ssh

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+# the repo has to be clone and updated properly so we can copy the changelogs so far.
+
+echo "Cloning the glusterfs-debian repo"
+git clone http://github.com/gluster/glusterfs-debian.git
```
This should be done over ssh if you want to commit, and https anyway, to avoid MITM. I am not sure github support http.

Shwetha-Acharya replied:
addressed

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+sudo pbuilder build ~/${os}-${flavor}-Glusterfs-${version}/build/glusterfs_${version}-${release}.dsc | tee build.log
+
+#move the packages to packages directory.
+mv /var/cache/pbuilder/result/glusterfs*${version}-${release}*.deb ~/${os}-${flavor}-Glusterfs-${version}/packages/
```
I am not sure jenkins used can erase file in /var/cache , so mv would fail.

Shwetha-Acharya replied:
@kalebskeithley Any suggestions on this?

mscherer replied:
I think it can safely replaced by "cp", at least.

Shwetha-Acharya replied:
done

mscherer replied:
Seems to still be mv :/

Shwetha-Acharya replied:
But these files should be removed to avoid eventual filling up of space, How is such case handled generally with jenkins? @mscherer

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+echo "Uploading the packages.."
+if [ "$os" == "ubuntu" ]; then
+        cd ..
+        dput ppa:gluster/glusterfs-${series} glusterfs_${version}-${os}1~${flavor}1_source.changes
```
Does it requires some authentication ? I do not see how it is done.

Shwetha-Acharya replied:
No, AIFIK. @kalebskeithley any more insights?

mscherer replied:
Well, if could be without authentication, but then, we need some kind of signature, cause I do not think anyone can push debian package for us, no ?

Shwetha-Acharya replied:
yes, we would need to follow https://help.launchpad.net/Packaging/PPA/Uploading to get required authentiacation

mscherer replied:
The documentation do not tell much, but if that use ssh, then we need to have the key available to the builder, which is not declared in the job yaml file. There is example here: https://github.com/gluster/build-jobs/blob/master/build-gluster-org/jobs/centos7-regression.yml#L64 for adding a credential

Shwetha-Acharya replied:
The  OpenPGP keys are used for signing as described [here](https://help.launchpad.net/YourAccount/ImportingYourPGPKey#:~:text=Publishing%20your%20key%20to%20a%20keyserver,-Step%201%3A%20Open&text=Choose%20the%20Sync%20button.,import%20key%20into%20into%20Launchpad.)
The above link also shows how a new key can be created and added to launcpad account.

$ gpg --list-keys on the machine rhs-vm-17.storage-dev.lab.eng.bos.redhat.com under the user glusterpackager, shows the already existing keys. (documented the same in the mojo doc draft: let me know if any more information is needed to be documented)

Shwetha-Acharya replied:
yes, we can store the key as secret in the job. We can add it in jenkins machine, I can help in getting it from  rhs-vm-17.storage-dev.lab.eng.bos.redhat.com

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+
+echo "Building source package.."
+cd ../glusterfs-${version}
+debuild -S -sa -k${debuild_key}
```
Where would the signing key stored ? (since -k requires that)

Shwetha-Acharya replied:
debuild_key variable contains the key.

mscherer replied:
I wasn't clear on my question. From what I understand, -k is the key identifier. But the actual private key is somewhere else (I think in ~/.gnupg), and so we need to discuss how that part is going to be managed.

Shwetha-Acharya replied:
we are initialising the values for debuild keys in https://github.com/gluster/build-jobs/pull/44/files/b32e0625d90b8c89432e753aab502d5e72b64b46#diff-02d2c32f5282eea2a7412f831420ae8c2f55ce0ed17671f6878e110c258b2e6bR29

The key creation is something we have maintained in a particular machine, whose access is not given to wider audience.

mscherer replied:
Then, where is that machine ?

Shwetha-Acharya replied:
Here is the draft of mojo doc that contains machine details: https://source.redhat.com/.motion/communitiesatredhat/crosscuttingco/pjmcop/prodev/study_grp/study_group_sign_up/A883E22D-0D6C-EB11-80F2-000D3A020FEB/.object

mscherer replied:
Page does not exist :/

Shwetha-Acharya replied:
Please checkout this link: https://source.redhat.com/.motion/communitiesatredhat/crosscuttingco/pjmcop/pjm_prodev/pjm_study_grp/pjm_study_group_sign_up/a883e22d-0d6c-eb11-80f2-000d3a020feb/.object

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+#copy the tar.gz file produced by the build to download.rht.gluster.org:/var/www/scratch
+scp $flavor-apt-amd64-$version.tgz gluster_ant@download.rht.gluster.org:/var/www/scratch
+
+ssh gluster_ant@download.rht.gluster.org
```
I do not get what this is supposed to do. Are you tryong to execute code on download.gluster.org (like, the subsequent commands ?)

Shwetha-Acharya replied:
yes.. the subsequent commands are to be executed in on download.gluster.org

mscherer replied:
Ok so this is not going to work as it is written. I would prefer if we had a script on download.rht.gluster.org, and execute it remotely (then we can lock down the execution with the ssh keys). But that's going to requires commit on more than 1 repo to be deployed.

Shwetha-Acharya replied:
added following script inside download.rht.gluster.org
`#!/bin/bash

series=$1
version=$2
os=$3
flavor=$4
latest_version=$5
latest_series=$6

cd /var/www/html/pub/gluster/glusterfs/$series/$version/$os/${flavor^}/amd64/apt
sudo tar xpf /var/www/scratch/$flavor*$version.tar.gz

#update latest inside the series you are building for
ln -snf /var/www/html/pub/gluster/glusterfs/$series/$version/$os/$flavor/amd64/apt /var/www/html/pub/gluster/glusterfs/$series/LATEST

#update latest inside the latest series
ln -snf /var/www/html/pub/gluster/glusterfs/$latest_series/$latest_version /var/www/html/pub/gluster/glusterfs/LATEST

#delete .tgz file from /var/www/scratch
rm -rf /var/www/scratch/$flavor*$version.tgz
`
We are not having seperate repo to track changes in download.rht.gluster.org

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+  sudo rm -rf /var/cache/pbuilder/base.tgz
+
+  #removing folders created while packaging
+  rm -rf ${os}-${flavor}-Glusterfs-${version}
```
Are we sure this will be run from the right directory ?

Shwetha-Acharya replied:
updated the path

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+sudo tar xpf /var/www/scratch/$flavor*$version.tar.gz
+
+#update latest inside the series you are building for
+ln -nf /var/www/html/pub/gluster/glusterfs/$series/$version/$os/$flavor/amd64/apt /var/www/html/pub/gluster/glusterfs/$series/LATEST
```
If we capitalize the flavor with ${flavor^}, shouldn't we do the same here ?

Also, isn't a "-s" missing here ? that's for a directory, and hardlink are not autorized for them, if I am not wrong.

Shwetha-Acharya replied:
fixed.

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+git checkout -b ${flavor}-${series}-local origin/${flavor}-glusterfs-${series}
+
+if [ "$os" == "ubuntu" ]; then
+	sed -i "1i glusterfs (${version}-${os}1~${flavor}1) ${flavor}; urgency=medium\n\n * GlusterFS ${version} GA\n\n -- GlusterFS GlusterFS deb packages <deb.packages@gluster.org>  `date +"%a, %d %b %Y %T %z"`\n" debian/changelog
```
I used dch in the past to modify the changelog, I think it would be more robust there than use sed.

Shwetha-Acharya replied:
fixed.

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+        dput ppa:gluster/glusterfs-${series} glusterfs_${version}-${os}1~${flavor}1_source.changes
+
+    echo "Done"
+    exit
```
so, why are we exiting here, since the package is not build yet ?

Shwetha-Acharya replied:
For Ubuntu, we are uploading the glusterfs_${version}-${os}1~${flavor}1_source.changes to [launchpad](https://launchpad.net/~gluster/+archive/ubuntu/glusterfs-7) Then the build process takes place on different architectures in the launchpad, which is automated by launchpad.

Shwetha-Acharya replied:
So, this exist is where, our part in building process can be marked done for ubuntu.

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+
+cd /var/www/repos/apt/debian/
+
+rm -rf pool/* dists/* db/*
```
Not sure jenkins can erase file there.

Shwetha-Acharya replied:
We can have a trail run to verify it.

mscherer replied:
Also, if we create file here, shouldn't it be cleaned with a trap, like the rest ?

Shwetha-Acharya replied:
yes, can be done

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,200 @@
+
+rm -rf pool/* dists/* db/*
+
+cp ~/conf.distributions/${series} conf/distributions
```
Where does thoses files come from ? Should we copy them to the builder ?

Shwetha-Acharya replied:
We can copy them to builder(where we run the automation) from the machines we are currently using for building.

Shwetha-Acharya committed:
Merge branch 'master' of https://github.com/Shwetha-Acharya/build-jobs

Shwetha-Acharya committed:
Add code to execute script inside download.rht.gluster.org
update mv to cp

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
Update the path for clean up
Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
Add loop instead of if

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
Change http to https during clone

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya commented:
Requesting next round of review @mscherer

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,169 @@
+  #removing folders created while packaging
+  rm -rf ~/${os}-${flavor}-Glusterfs-${version}
+}
+trap finish EXIT
```
Traps should be placed at the start of the script. Otherwise, it be used only when the script is over, which is not useful.

Shwetha-Acharya replied:
fixed

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,169 @@
+#copy the tar.gz file produced by the build to download.rht.gluster.org:/var/www/scratch
+scp $flavor-apt-amd64-$version.tgz gluster_ant@download.rht.gluster.org:/var/www/scratch
+
+ssh gluster_ant@download.rht.gluster.org /var/www/html/pub/gluster/unpacking-script.sh series version os flavor latest_version latest_series
```
Shouldn't it be "$version", etc ?

Shwetha-Acharya replied:
fixed

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,169 @@
+
+#move the packages to packages directory.
+cp /var/cache/pbuilder/result/glusterfs*${version}-${release}*.deb ~/${os}-${flavor}-Glusterfs-${version}/packages/
+rm -rf /var/cache/pbuilder/result/glusterfs*${version}-${release}*.deb
```
Why not mv instead of cp + rm ?

Shwetha-Acharya replied:
done

mscherer commented on build-gluster-org/scripts/generic_package.sh:
```diff
@@ -0,0 +1,169 @@
+cp /var/cache/pbuilder/result/glusterfs*${version}-${release}*.deb ~/${os}-${flavor}-Glusterfs-${version}/packages/
+rm -rf /var/cache/pbuilder/result/glusterfs*${version}-${release}*.deb
+
+if [ "$flavor" != "stretch" ]; then
```
So, why is stretch special here ?

Shwetha-Acharya replied:
It is because /var/cache/pbuilder/result/libg*${version}-${release}*.deb are not created only in stretch. In buster and bullseye, they will be created

Shwetha-Acharya committed:
address issue with trap trap
replace cp+rm with mv
add misisng $

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
use ssh for cloning from github

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
refactor loops

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
remove comment with respect to series 5

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya commented:
@mscherer all the requested change sets are uploaded.

mscherer commented:
(and conflicts)

Shwetha-Acharya committed:
modify shell path in debian-ubuntu-package.yml
according to the actual loaction of the script

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
Alter path with respect to ${WORKSPACE}

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
Create packages inside the directory wrt ${WORKSPACE}

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya commented:
> So, since the review is getting a bit messy with comments, the blocking part is the gpg key secret integration for now.

Action plan to resolve gpg key integration part: https://docs.google.com/document/d/1KRuzMC9zPD6K-5k03_5vUGjWrR0kUFiN2UhaYm18phQ/edit?usp=sharing

Also as agreed earlier, infra team is upposed to create the following users for these scripts to work:
1. user named gluster_ant  in the build machine, who has permission to access download.rht.gluster.org , so that 
    built packages can be pushed to download.rht.gluster.org
   
2. github user so that script can push the changelogs to respective github reposotories.

deepshikhaaa commented on build-gluster-org/jobs/debian-ubuntu-package.yml:
```diff
@@ -0,0 +1,57 @@
+- job:
+    name: debian-package-builder
+    node: master
```
Use the label `debian10` for this job. We have a machine existing on jenkins https://build.gluster.org/computer/builder-deb10-1.int.rht.gluster.org/

Shwetha-Acharya replied:
@deepshikhaaa done!

Shwetha-Acharya replied:
@deepshikhaaa please validate the liburing-devel support as well.

Shwetha-Acharya committed:
Add debian10 as the label for the node inorder to use
https://build.gluster.org/computer/builder-deb10-1.int.rht.gluster.org/

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya commented:
We need a machaine with kernel 5.1 or above inorder to enable io-uring support with the builds. Please let me know which node can be used.

As mentioned earlier there is an action plan to include  [Action plan](https://docs.google.com/document/d/1KRuzMC9zPD6K-5k03_5vUGjWrR0kUFiN2UhaYm18phQ/edit?usp=sharing) to resolve gpg key integration part.

@mscherer @deepshikhaaa can we take this forward?
Going forward it is very much essential to keep this packaging and building part automated.

mscherer commented:
We only have Debian 10 hosts, not Ubuntu one. I can try to spin a Ubuntu one but that would be the 1st we have, so I am unsure how long it will take.

mscherer commented:
Ok so the Ubuntu installer do not work (and I do not understand why yet, there is no obvious error message). But so, do we need the kernel to be up to date for package building ? Given this will use a pbuilder, the headers would be there and nothing more should be needed, unless I am missing something ?

mscherer commented:
So the Ubuntu installer didn't work because:
- the firewall blocked the network (each time, i forget)
- some weird ansible issue with variable (so it was installing centos), I fixed
- then ubuntu changed the installer and so our RHEL 7 host is too old to detect that 

So I am doing 18.04 and upgrade manually.

Shwetha-Acharya commented:
> Ok so the Ubuntu installer do not work (and I do not understand why yet, there is no obvious error message). But so, do we need the kernel to be up to date for package building ? Given this will use a pbuilder, the headers would be there and nothing more should be needed, unless I am missing something ?

Right, in this script the latest chroot will be created.  We would require a ubuntu mechine as discussed offline as it is tested in ubuntu.

mscherer commented:
But testing is done on a different job than the packaging one, no ? I still do not get the issue, at what point is the deb packages tested, and how ?

mscherer commented:
So we have 20.04 builder now: https://build.gluster.org/computer/builder-ubuntu2004-1.int.rht.gluster.org/

Shwetha-Acharya committed:
update the node label to ubuntu2004
(https://build.gluster.org/computer/builder-ubuntu2004-1.int.rht.gluster.org/)

Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>
